### DISCUSSION TITLE
Does Jena contains a sparql federated query engine that doesn't use the SERVICE keyword?

### DISCUSSION THREAD

**Discussion by chenkuncloud** [apache/jena#1839] [Q&A] [answered]
Hi, does `Jena` contains a sparql federated query engine that doesn't use the `SERVICE` keyword? 

Such as [FedX of RDF4J](https://rdf4j.org/documentation/programming/federation/)
```
Repository repository = FedXFactory.newFederation()
	.withSparqlEndpoint("http://dbpedia.org/sparql")
	.withSparqlEndpoint("https://query.wikidata.org/sparql")
	.create();

try (RepositoryConnection conn = repository.getConnection()) {

	String query =
		"PREFIX wd: <http://www.wikidata.org/entity/> "
		+ "PREFIX wdt: <http://www.wikidata.org/prop/direct/> "
		+ "SELECT * WHERE { "
		+ " ?country a <http://dbpedia.org/class/yago/WikicatMemberStatesOfTheEuropeanUnion> ."
		+ " ?country <http://www.w3.org/2002/07/owl#sameAs> ?countrySameAs . "
		+ " ?countrySameAs wdt:P2131 ?gdp ."
		+ "}";

	TupleQuery tq = conn.prepareTupleQuery(query);
	try (TupleQueryResult tqRes = tq.evaluate()) {

		int count = 0;
		while (tqRes.hasNext()) {
			BindingSet b = tqRes.next();
			System.out.println(b);
			count++;
		}

		System.out.println("Results: " + count);
	}
}

repository.shutDown();
```

**Comment by afs** [comment]
Not as such. Jena does have a subsystem for working with remote queries, controlling caching and the execution. See [Service Enhancer](https://jena.apache.org/documentation/query/service_enhancer.html).

**Reply by chenkuncloud** [reply]
Hi @afs , I'm sorry that I may not be specific enough, but I actually want to use `SPARQL` without the `SERVICE` keyword to query `multiple SPARQL endpoints`, for example:
- Multiple SPARQL endpoints: http://linkedgeodata.org/sparql, http://linkedgeodata.org/geonames
- SPARQL:
```text
SELECT * WHERE {
  <http://dbpedia.org/resource/Yao_Ming> <http://dbpedia.org/ontology/birthPlace> ?dbpediaBirthPlace .
  ?dbpediaBirthPlace rdfs:label ?birthPlaceLabel .
  ?geoBirthPlace rdfs:label ?birthPlaceLabel .
  ?geoBirthPlace <http://www.w3.org/2003/01/geo/wgs84_pos#long> ?long .
  ?geoBirthPlace <http://www.w3.org/2003/01/geo/wgs84_pos#lat> ?lat .
} LIMIT 1
```

**Reply by afs** [reply, marked as answer]
That as understood. There isn't some in the Jena releases. [Service Enhancer](https://jena.apache.org/documentation/query/service_enhancer.html) and a single outer SERVICE to trigger that subsystem to look at he pattern may have some capability. Maybe someone will comment.

**Reply by chenkuncloud** [reply]
Thanks @afs , by the way, will jena support this feature in the future?

**Reply by afs** [reply]
Contributions welcome!
